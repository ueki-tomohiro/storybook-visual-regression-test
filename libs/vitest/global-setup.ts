const setup = () => {
  process.env.TZ = "Asia/Tokyo";
};

export default setup;
