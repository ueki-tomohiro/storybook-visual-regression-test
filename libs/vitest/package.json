{
  "name": "@demo-libs/vitest",
  "version": "0.0.3",
  "private": true,
  "description": "Demo Libs Vitest",
  "license": "MIT",
  "type": "module",
  "main": "index.js",
  "devDependencies": {
    "@juggle/resize-observer": "^3.4.0",
    "@testing-library/react": "^16.0.1",
    "@testing-library/user-event": "^14.5.2",
    "@vitejs/plugin-react": "^4.3.1",
    "@vitejs/plugin-react-swc": "^3.7.0",
    "@vitest/coverage-v8": "^2.0.5",
    "canvas": "^2.11.2",
    "jsdom": "^25.0.0",
    "vite-tsconfig-paths": "^5.0.1",
    "vitest": "^2.0.5"
  },
  "engines": {
    "node": "^20",
    "yarn": ">=1"
  },
  "volta": {
    "node": "20.0.0",
    "yarn": "1.22.19"
  }
}
