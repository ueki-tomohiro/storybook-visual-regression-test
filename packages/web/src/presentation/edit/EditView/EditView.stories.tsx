import { getGetTodoMock } from "@demo/api/lib/demo/todo/todo.msw";
import { action } from "@storybook/addon-actions";

import { EditView, EditViewProps } from "./EditView";

const defaultArgs: EditViewProps = {
  todo: {
    todo_id: 90856,
    description: "turquoise",
    end_date: "2022-12-29T16:14:31+09:00",
    completed: true,
  },
  updateTodo: action("updateTodo"),
  isLoading: false,
};

export default {
  title: "views/EditView",
  component: EditView,
};

export const Default = { args: defaultArgs };

export const Loading = { args: { ...defaultArgs, isLoading: true } };

export const Fake = {
  args: {
    ...defaultArgs,
    todo: getGetTodoMock(),
  },
};
