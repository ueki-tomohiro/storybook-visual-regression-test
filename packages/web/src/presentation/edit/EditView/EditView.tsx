import { Todo, TodoUpdatable } from "@demo/api/lib/demo/model";
import { Button } from "@demo/ui/components/Button";
import { Checkbox } from "@demo/ui/components/Checkbox";
import { InputField } from "@demo/ui/components/InputField";
import { yupResolver } from "@hookform/resolvers/yup";
import React from "react";
import { SubmitHandler, useForm } from "react-hook-form";
import * as Yup from "yup";

export type EditViewProps = {
  todo: Todo;
  updateTodo: (args: TodoUpdatable) => void;
  isLoading?: boolean;
};

export const EditView: React.FC<EditViewProps> = ({ todo, updateTodo, isLoading }) => {
  const resolver = yupResolver(
    Yup.object().shape({
      description: Yup.string().required("詳細を入力してください"),
      end_date: Yup.string().typeError("完了日を入力してください").required("完了日を入力してください"),
      completed: Yup.boolean().required(),
    })
  );
  const {
    handleSubmit,
    register,
    formState: { errors, isSubmitting },
  } = useForm<TodoUpdatable>({
    defaultValues: {
      description: todo.description,
      end_date: todo.end_date,
      completed: todo.completed,
    },
    resolver,
  });

  const onSubmit: SubmitHandler<TodoUpdatable> = (values) => {
    updateTodo(values);
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <InputField
        {...register("description")}
        label="詳細"
        disabled={isSubmitting || isLoading}
        error={errors.description?.message}
      />
      <InputField
        {...register("end_date")}
        label="完了日"
        type="date"
        disabled={isSubmitting || isLoading}
        error={errors.end_date?.message}
      />
      <Checkbox {...register("completed")} label="完了" disabled={isSubmitting || isLoading} />
      <Button variant="primary" type="submit" disabled={isSubmitting || isLoading} isLoading={isLoading}>
        登録
      </Button>
    </form>
  );
};
