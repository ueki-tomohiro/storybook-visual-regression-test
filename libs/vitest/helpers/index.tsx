import { composeStories, composeStory } from "@storybook/react";
import type { Args } from "@storybook/types";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { cleanup, render } from "@testing-library/react";
import dayjs from "dayjs";
import noop from "lodash/noop";
import React, { ReactNode } from "react";
import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";

const date = dayjs("2020-12-04T10:00:00+09:00").toDate();

export const renderFromStoriesTest = (stories: Parameters<typeof composeStories<any>>[0]) => {
  const componentTitle = stories.default.title as string;
  const defaultArgs = (stories.default && stories.default.args) || {};

  describe(componentTitle, () => {
    Object.entries(composeStories(stories)).forEach(([storyName, Story]: [string, any]) => {
      if (storyName === "__esModule") return;
      if (storyName === "default") return;
      if (storyName.match(/Fake/)) return;

      beforeEach(() => {
        vi.setSystemTime(date);
        Object.defineProperty(HTMLMediaElement.prototype, "muted", {
          set: noop,
        });
        const scrollTo = vi.fn();
        vi.spyOn(window, "scrollTo").mockImplementation(scrollTo);
      });
      afterEach(() => {
        vi.useRealTimers();
        cleanup();
      });

      describe(`"${storyName}" story`, () => {
        test("renders without crashing", () => {
          const { container, unmount } = render(<Story {...defaultArgs} {...(Story.args || {})} />);
          expect(container).toMatchSnapshot();
          unmount();
        });
      });
    });
  });
};

export const interactionTestFromStory = <T extends Args>({
  testName,
  story,
  annotations,
  args,
  callback,
}: {
  testName?: string;
  story: Parameters<typeof composeStory<T>>[0];
  annotations: Parameters<typeof composeStory<T>>[1];
  args?: Partial<T>;
  callback: (args: {
    container: HTMLElement;
    rerender: (ui: React.ReactElement) => void;
    Story: ReturnType<typeof composeStory<T>>;
  }) => Promise<void>;
}) => {
  const Story = composeStory<T>(story, { ...annotations, args: { ...annotations.args, ...args } });

  beforeEach(() => {
    vi.setSystemTime(date);
  });
  afterEach(() => {
    vi.useRealTimers();
    cleanup();
  });

  describe(`"${annotations.title}" story`, () => {
    test(`interaction test ${testName || ""}`, async () => {
      const { container, rerender } = render(<Story {...args} />);
      await callback({ container, rerender, Story });
    });
  });
};

export const createQueryWrapper = () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        staleTime: Infinity,
      },
    },
  });

  const QueryWrapper = ({ children }: { children: ReactNode }) => (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );

  return {
    queryClient,
    QueryWrapper,
  };
};
