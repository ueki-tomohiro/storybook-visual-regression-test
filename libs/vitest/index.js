import react from "@vitejs/plugin-react-swc";
import tsconfigPaths from "vite-tsconfig-paths";
import { defineConfig } from "vitest/config";

function stubAssetImport() {
  return {
    name: "stub-asset-import",
    transform(_code, id) {
      if (/(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$/.test(id))
        return {
          code: "export default 'test-file-stub'",
        };
    },
  };
}

/** @type { import('vite').UserConfig } */
export default defineConfig({
  plugins: [react(), tsconfigPaths(), stubAssetImport()],
  test: {
    globalSetup: "@demo-libs/vitest/global-setup.ts",
    root: ".",
    environment: "jsdom",
    setupFiles: ["@demo-libs/vitest/vitest.setup.mts"],
    testTransformMode: {
      ssr: ["**/*"],
    },
    testTimeout: 10000,
    css: { modules: { classNameStrategy: "non-scoped" }, include: /.*/ },
    poolOptions: {
      threads: {
        singleThread: true,
      },
    },
    reporters: ["default", "hanging-process"],
  },
});
