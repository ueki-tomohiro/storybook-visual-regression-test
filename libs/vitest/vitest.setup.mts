import "@testing-library/jest-dom/vitest";

import { configure } from "@testing-library/react";
export * from "./helpers";

configure({
  testIdAttribute: "data-testid",
});
