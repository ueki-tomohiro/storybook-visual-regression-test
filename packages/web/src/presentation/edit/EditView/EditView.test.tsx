import { renderFromStories } from "@demo-libs/vitest/helpers";
import { describe } from "vitest";

import * as stories from "./EditView.stories";

describe(stories.default.title, () => {
  renderFromStories(stories);
});
