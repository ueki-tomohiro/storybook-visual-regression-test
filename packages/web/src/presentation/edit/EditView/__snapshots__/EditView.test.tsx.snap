// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`views/EditView renders Default 1`] = `
<body>
  <div>
    <form>
      <div
        class="relative w-full"
      >
        <fieldset
          class="w-full h-12 mb-6 rounded border border-solid border-tertiary focus-within:border-secondary [&_label]:[&:not(:focus-within)]:text-tertiary [&_label]:focus-within:text-secondary"
        >
          <legend
            class="min-h-3 ml-2 flex px-0.5"
          >
            <label
              class="flex text-xs"
            >
              詳細
            </label>
          </legend>
          <div
            class="absolute inset-0 flex items-centerpt-1.5"
          >
            <input
              class="bg-reset h-11 w-full px-2 text-tertiary focus:text-secondary focus:outline-none"
              name="description"
              type="text"
              value=""
            />
          </div>
          <div
            class="-z-1 absolute inset-0"
          />
        </fieldset>
      </div>
      <div
        class="relative w-full"
      >
        <fieldset
          class="w-full h-12 mb-6 rounded border border-solid border-tertiary focus-within:border-secondary [&_label]:[&:not(:focus-within)]:text-tertiary [&_label]:focus-within:text-secondary"
        >
          <legend
            class="min-h-3 ml-2 flex px-0.5"
          >
            <label
              class="flex text-xs"
            >
              完了日
            </label>
          </legend>
          <div
            class="absolute inset-0 flex items-centerpt-1.5"
          >
            <input
              class="bg-reset h-11 w-full px-2 text-tertiary focus:text-secondary focus:outline-none"
              name="end_date"
              type="date"
              value=""
            />
          </div>
          <div
            class="-z-1 absolute inset-0"
          />
        </fieldset>
      </div>
      <label
        class="mb-4 flex"
      >
        <div
          class="mr-2 grow-0"
        >
          <input
            class="mr-2 scale-150 accent-primary"
            name="completed"
            type="checkbox"
          />
        </div>
        <div
          class="grow font-bold"
        >
          完了
        </div>
      </label>
      <button
        class="mx-auto max-w-3xl text-white bg-primary rounded px-4 py-2 disabled:bg-tertiary"
        type="submit"
      >
        登録
      </button>
    </form>
  </div>
</body>
`;

exports[`views/EditView renders Loading 1`] = `
<body>
  <div>
    <form>
      <div
        class="relative w-full"
      >
        <fieldset
          class="w-full h-12 mb-6 rounded border border-solid border-tertiary focus-within:border-secondary [&_label]:[&:not(:focus-within)]:text-tertiary [&_label]:focus-within:text-secondary"
        >
          <legend
            class="min-h-3 ml-2 flex px-0.5"
          >
            <label
              class="flex text-xs"
            >
              詳細
            </label>
          </legend>
          <div
            class="absolute inset-0 flex items-centerpt-1.5"
          >
            <input
              class="bg-reset h-11 w-full px-2 text-tertiary focus:text-secondary focus:outline-none"
              disabled=""
              name="description"
              type="text"
              value=""
            />
          </div>
          <div
            class="-z-1 absolute inset-0"
          />
        </fieldset>
      </div>
      <div
        class="relative w-full"
      >
        <fieldset
          class="w-full h-12 mb-6 rounded border border-solid border-tertiary focus-within:border-secondary [&_label]:[&:not(:focus-within)]:text-tertiary [&_label]:focus-within:text-secondary"
        >
          <legend
            class="min-h-3 ml-2 flex px-0.5"
          >
            <label
              class="flex text-xs"
            >
              完了日
            </label>
          </legend>
          <div
            class="absolute inset-0 flex items-centerpt-1.5"
          >
            <input
              class="bg-reset h-11 w-full px-2 text-tertiary focus:text-secondary focus:outline-none"
              disabled=""
              name="end_date"
              type="date"
              value=""
            />
          </div>
          <div
            class="-z-1 absolute inset-0"
          />
        </fieldset>
      </div>
      <label
        class="mb-4 flex"
      >
        <div
          class="mr-2 grow-0"
        >
          <input
            class="mr-2 scale-150 accent-primary"
            disabled=""
            name="completed"
            type="checkbox"
          />
        </div>
        <div
          class="grow font-bold"
        >
          完了
        </div>
      </label>
      <button
        class="mx-auto max-w-3xl text-white bg-primary rounded px-4 py-2 disabled:bg-tertiary"
        disabled=""
        type="submit"
      >
        <span
          class="material-symbols-outlined animate-rotate"
        >
          refresh
        </span>
      </button>
    </form>
  </div>
</body>
`;
